iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -324,6 +326,249 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
+    auto rhsShapeType = op.rhs().getType().dyn_cast_or_null<ShapedType>();
+    auto resultType = op.getResult().getType().dyn_cast_or_null<ShapedType>();
+
+    if (!lhsShapeType || !rhsShapeType || !resultType) return failure();
+    if (!lhsShapeType.hasStaticShape() || !rhsShapeType.hasStaticShape())
+      return failure();
+    if (resultType.getRank() <= 3) return failure();
+
+    mhlo::DotDimensionNumbers dimNumbers = op.dot_dimension_numbers();
+    auto lhsBatchingDims = llvm::to_vector<4>(
+        llvm::map_range(dimNumbers.lhs_batching_dimensions(),
+                        [](APInt v) { return v.getSExtValue(); }));
+    auto rhsBatchingDims = llvm::to_vector<4>(
+        llvm::map_range(dimNumbers.rhs_batching_dimensions(),
+                        [](APInt v) { return v.getSExtValue(); }));
+    auto lhsContractingDims = llvm::to_vector<4>(
+        llvm::map_range(dimNumbers.lhs_contracting_dimensions(),
+                        [](APInt v) { return v.getSExtValue(); }));
+    auto rhsContractingDims = llvm::to_vector<4>(
+        llvm::map_range(dimNumbers.rhs_contracting_dimensions(),
+                        [](APInt v) { return v.getSExtValue(); }));
+
+    if (lhsBatchingDims.empty() || rhsBatchingDims.empty()) return failure();
+
+    llvm::sort(lhsBatchingDims);
+    llvm::sort(lhsContractingDims);
+    llvm::sort(rhsBatchingDims);
+    llvm::sort(rhsContractingDims);
+
+    auto isConsecutive = [](ArrayRef<int64_t> array) {
+      for (int i = 1; i < array.size(); ++i) {
+        if (array[i] - array[i - 1] != 1) return false;
+      }
+      return true;
+    };
+
+    auto isDomainSplit = [](ArrayRef<int64_t> shape,
+                            ArrayRef<int64_t> batchingDims,
+                            ArrayRef<int64_t> contractingDims) {
+      // Batching and contracting are contiguous.
+      if ((contractingDims.front() - batchingDims.back()) == 1) return false;
+      // Contracting dims are inner most.
+      if (contractingDims.back() == (shape.size() - 1)) return false;
+      return true;
+    };
+
+    if (!isConsecutive(lhsBatchingDims) || !isConsecutive(lhsContractingDims) ||
+        !isConsecutive(rhsBatchingDims) || !isConsecutive(rhsContractingDims))
+      return failure();
+
+    if (isDomainSplit(lhsShapeType.getShape(), lhsBatchingDims,
+                      lhsContractingDims) ||
+        isDomainSplit(rhsShapeType.getShape(), rhsBatchingDims,
+                      rhsContractingDims))
+      return failure();
+
+    // Collabsing shape into a rank-3 tensor, returns newCollabsedShape
+    // contraction and parallel dim indices.
+    auto computeCollabsedShape = [](ArrayRef<int64_t> shape,
+                                    ArrayRef<int64_t> batchingDims,
+                                    ArrayRef<int64_t> contractingDims) {
+      auto newRank =
+          shape.size() - batchingDims.size() - contractingDims.size() + 2;
+      llvm::SmallVector<int64_t, 4> newShape(newRank);
```

Comment:
Move it to below where it's used.
And maybe can write like: llvm::SmallVector<...> newShape = { ... };